### PR TITLE
[APCu] A new polyfill for the legacy APC users

### PR DESCRIPTION
**Pull request by nicolas-grekas**
Now that we know that `apc_*` functions are legacy and `apcu_*` is the only way in PHP7, I propose to move forward with this polyfill. Let's replace all usages of `apc_*` in Symfony and elsewhere by their `apcu_*` equivalent and add a deps on this new polyfill to cope with legacy `apc` users.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
ping @dunglas @wouterj @ogizanagi

File: src/Apcu/Apcu.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        $entry = apcu_fetch($key, $success);
+
+        if ($success) {
+            return $entry;
+        }
+
+        $h = fopen(__FILE__, 'r');
+        flock($h, LOCK_EX);
```
Comment (dunglas) [inline src/Apcu/Apcu.php]: Is it sufficient to avoid race conditions?
What if the file is already being executed by another thread before the call to `flock`? What's the behavior when the opcache caching is on?

Maybe can you just throw a `NotSupportedException` for this specific function?

Comment (nicolas-grekas) [inline src/Apcu/Apcu.php]: > Is it sufficient to avoid race conditions?
I think so
> What if the file is already being executed by another thread
locking happens at the kernel level and is share among threads/processes
> throw a NotSupportedException
as last resort  but for now I see no issue here

**Comment by dunglas** [comment]
Btw, APC is not used directly in Symfony except in the `ApcClassLoader` (https://github.com/symfony/symfony/search?utf8=%E2%9C%93&q=apc_fetch&type=Code).

Most APC calls are done by the Doctrine Cache library.

File: src/Apcu/Apcu.php (hunk 1 above)
Comment (dunglas) [inline src/Apcu/Apcu.php]: Nervermind it should be fine. At least it's ok for the case I described before.

File: src/Apcu/composer.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=5.3.3"
+    },
+    "autoload": {
+        "files": [ "bootstrap.php" ],
+        "classmap": [ "Resources/stubs" ]
```
Comment (stof) [inline src/Apcu/composer.json]: missing autoloading for the ``Symfony\Polyfill\Apcu\Apcu`` class

File: src/Apcu/Apcu.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * @internal
+ */
+final class Apcu
+{
+    public static function apcu_entry($key, $generator, $ttl = 0)
+    {
+        if (!is_scalar($key)) {
+            throw new \InvalidArgumentException(sprintf('apcu_entry() expects parameter 1 to be string, %s given', __FUNCTION__, gettype($value)));
```
Comment (stof) [inline src/Apcu/Apcu.php]: there is no placeholder for ``__FUNCTION__``

Comment (nicolas-grekas) [inline src/Apcu/Apcu.php]: fixed

File: src/Apcu/composer.json (hunk 2 above)
Comment (stof) [inline src/Apcu/composer.json]: this needs to be fixed

Comment (nicolas-grekas) [inline src/Apcu/composer.json]: good catch, fixed

**Comment by nicolas-grekas** [comment]
Ping @symfony/deciders

**Comment by dunglas** [comment]
:+1:

File: src/Apcu/LICENSE [hunk 4]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2014-2015 Fabien Potencier
```
Comment (Tobion) [inline src/Apcu/LICENSE]: 2016 only

File: .travis.yml [hunk 5]
```diff
@@ -14,12 +14,16 @@ matrix:
...
 before_install:
     - composer self-update;
-    - if [[ "$TRAVIS_PHP_VERSION" != "hhvm" ]]; then phpenv config-rm xdebug.ini; fi;
-    - if [[ "$TRAVIS_PHP_VERSION" != "hhvm" ]]; then echo "extension = ldap.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini; fi;
-    - if [[ "$TRAVIS_PHP_VERSION" != "hhvm" ]]; then php -i; fi;
+    - if [[ $TRAVIS_PHP_VERSION != hhvm ]]; then INI_FILE=~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini; else INI_FILE=/etc/hhvm/php.ini; fi;
+    - echo memory_limit = -1 >> $INI_FILE
+    - echo session.gc_probability = 0 >> $INI_FILE
+    - if [[ $TRAVIS_PHP_VERSION != hhvm ]]; then phpenv config-rm xdebug.ini; fi;
+    - if [[ $TRAVIS_PHP_VERSION = 5.* ]]; then (echo yes | pecl install -f apcu-4.0.10 && echo apc.enable_cli = 1 >> $INI_FILE) || echo "Let's continue without apcu extension"; fi;
```
Comment (Tobion) [inline .travis.yml]: why `apc.enable_cli` here and in `phpunit.xml.dist`

Comment (nicolas-grekas) [inline .travis.yml]: I removed it from phpunit.xml.dist since it can't be "ini_setted"

File: src/Apcu/LICENSE (hunk 4 above)
Comment (nicolas-grekas) [inline src/Apcu/LICENSE]: fixed

File: tests/Apcu/ApcuTest.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        $this->assertSame(array(__CLASS__), array_keys($entries));
+        $this->assertSame(__CLASS__, $entries[__CLASS__]['key']);
+        $this->assertSame(456, $entries[__CLASS__]['value']);
+    }
+
+    public function testApcuEntry()
+    {
+        apcu_delete(__CLASS__);
+
+        $gen = function ($key) use (&$ok) {
```
Comment (Tobion) [inline tests/Apcu/ApcuTest.php]: shoulnd't `$ok` be intitialized first?

Comment (nicolas-grekas) [inline tests/Apcu/ApcuTest.php]: It could but that would be exactly the same, references are automatically assigned to null

File: src/Apcu/Apcu.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * @internal
+ */
+final class Apcu
+{
+    public static function apcu_entry($key, $generator, $ttl = 0)
+    {
+        if (!is_scalar($key)) {
+            throw new \InvalidArgumentException(sprintf('apcu_entry() expects parameter 1 to be string, %s given', gettype($value)));
```
Comment (Tobion) [inline src/Apcu/Apcu.php]: this is not the same behavior as the real apcu_entry` method. it does not throw an exception and it also does not seem to generate a warning either from what I tested. It just return null.

**Comment by Tobion** [comment]
The doc about `apcu_entry http://php.net/manual/en/function.apcu-entry.php says

> In addition, it prohibits the concurrent execution of any other APCu functions, since they will acquire the same lock.

This does not happen in the current polyfill implementation. This means you can still have race conditions if another process modifies the data while you are working with it inside the generator callable.

File: tests/Apcu/ApcuTest.php [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Polyfill\Tests\Apcu;
+
+/**
+ * @requires extension apc
```
Comment (Tobion) [inline tests/Apcu/ApcuTest.php]: shouldn't these tests run against both apc and apcu to ensure they work the same. now it's only executed against apc.

File: src/Apcu/Apcu.php (hunk 7 above)
Comment (nicolas-grekas) [inline src/Apcu/Apcu.php]: Thanks for testing this, I will add a test then

File: tests/Apcu/ApcuTest.php (hunk 8 above)
Comment (nicolas-grekas) [inline tests/Apcu/ApcuTest.php]: Apcu provides apc, it returns true to extension_loaded('apc')

**Comment by nicolas-grekas** [comment]
@tobion then we should acquire the lock in every apc function call, I'll add it tomorrow.

**Comment by nicolas-grekas** [comment]
@Tobion I removed the polyfilling of `apcu_entry`: too many edge cases with locking, and not required for moving forward.

**Comment by fabpot** [comment]
:+1:

File: tests/Apcu/ApcuTest.php (hunk 8 above)
Comment (Tobion) [inline tests/Apcu/ApcuTest.php]: Hm not for me using APCu 5.1.2 on Windows

```
$ php --version
PHP 7.0.1 (cli) (built: Dec 16 2015 13:32:58) ( NTS )
$ php -d "apc.enable_cli=1" -r "var_dump(extension_loaded('apc'));"
bool(false)
$ php -d "apc.enable_cli=1" -r "var_dump(extension_loaded('apcu'));"
bool(true)
```

Comment (nicolas-grekas) [inline tests/Apcu/ApcuTest.php]: so, it looks like you don't have the apcu bc layer enabled, but that doesn't change my point: this is tested properly on travis, against apcu + apc-br-from-apcu, which means there is nothing to fix, the case is covered

Comment (nicolas-grekas) [inline tests/Apcu/ApcuTest.php]: OH, and btw, this polyfill makes no sense on PHP 7 (where apc has never been released and apcu is the only avail thing)
